src/platform_impl/linux/x11/event_processor.rs

```diff
@@ -687,7 +709,18 @@ impl<T: 'static> EventProcessor<T> {
                         let window_id = mkwid(xev.event);
                         let new_cursor_pos = (xev.event_x, xev.event_y);
 
-                        let modifiers = ModifiersState::from(xev.mods);
+                        let modifiers = ModifiersState::from_x11(&xev.mods);
+
+                        if let Some(modifiers) =
+                            self.device_mod_state.update_state(&modifiers, None)
```

Comment:
I feel like this check and later modifiers changed event send should be extracted to an `inline` function, since they are mostly identical (I'm about XI_Motion/XI_ButtonPress/ffi::KeyPress) and share the same logic.

Reply:
Fair point. I've replaced the repeated `if let` with a macro invocation, using a macro defined at the top of the function. As noted in the comment, a `&mut self` method can't be used because `&self.target` is borrowed for the entire body of the function.